ethexe/contracts/src/Router.sol

```diff
@@ -207,6 +229,46 @@ contract Router is IRouter, OwnableUpgradeable, ReentrancyGuardTransient {
     }
 
     // # Validators calls.
+
+    /// @dev Set validators for the next era.
+    function commitValidators(Gear.ValidatorsCommitment calldata commitment, bytes[] calldata signatures) external {
+        Storage storage router = _router();
+
+        uint256 currentEraIndex = (block.timestamp - router.genesisBlock.timestamp) / router.durations.era;
+
+        require(commitment.eraIndex == currentEraIndex + 1, "commitment era index is not next era index");
+
+        uint256 nextEraStart = router.genesisBlock.timestamp + router.durations.era * commitment.eraIndex;
+        require(block.timestamp >= nextEraStart - router.durations.election, "election is not yet started");
+
+        bool useValidators0 = Gear.currentEraValidatorsStoredInValidators1(router);
+        if (useValidators0) {
+            require(
+                router.validationSettings.validators0.useFromTimestamp < block.timestamp,
+                "looks like validators for next era are already set"
+            );
+        } else {
+            require(
+                router.validationSettings.validators1.useFromTimestamp < block.timestamp,
+                "looks like validators for next era are already set"
+            );
+        }
+
+        bytes32 commitmentHash = keccak256(bytes.concat(Gear.validatorsCommitmentHash(commitment)));
```

Comment:
```suggestion
        bytes32 commitmentHash = keccak256(abi.encodePacked(Gear.validatorsCommitmentHash(commitment)));
```

Reply:
ah, ok that's because elsewhere we do `keccak256(bytes memory)` instead of `keccak256(bytes32)`. then write it like this:
```solidity
bytes32 commitmentHash = Gear.validatorsCommitmentHash(commitment);
require(
    Gear.validateSignatures(router, keccak256(abi.encodePacked(commitmentHash)), signatures),
    "next era validators signatures verification failed"
);
```

Reply:
done